interop-testing/src/test/java/io/grpc/testing/integration/TlsTest.java

```diff
@@ -0,0 +1,253 @@
+    int port = TestUtils.pickUnusedPort();
+
+    // Create & start a server.
+    File serverCertFile = TestUtils.loadCert("localhost_server.pem");
+    File serverPrivateKeyFile = TestUtils.loadCert("localhost_server.key");
+    X509Certificate[] serverTrustedCaCerts = {
+      TestUtils.loadX509Cert("ca.pem")
+    };
+    Server server = serverBuilder(port, serverCertFile, serverPrivateKeyFile, serverTrustedCaCerts)
+        .addService(EchoServiceGrpc.bindService(new DummyEchoRpcService()))
+        .build()
+        .start();
+
+    try {
+      // Create a client.
+      File clientCertFile = TestUtils.loadCert("client.pem");
+      File clientPrivateKeyFile = TestUtils.loadCert("client.key");
+      X509Certificate[] clientTrustedCaCerts = {
+        TestUtils.loadX509Cert("ca.pem")
+      };
+      ManagedChannel channel = clientChannel("localhost", port, clientCertFile,
+                                             clientPrivateKeyFile, clientTrustedCaCerts);
+      EchoServiceBlockingStub client = EchoServiceGrpc.newBlockingStub(channel);
+
+      // Send an actual request, via the full GRPC & network stack, and check that a proper
+      // response comes back.
+      EchoRequest request = EchoRequest.newBuilder()
+          .setText("dummy text")
+          .build();
+      EchoResponse response = client.echo(request);
+      assertEquals("Request said: dummy text", response.getText());
+    } finally {
+      server.shutdown();
+    }
+  }
+
+
+  /**
+   * Tests that a server configured to require client authentication refuses to accept connections
+   * from a client that has an untrusted certificate.
+   */
+  @Test
+  public void serverRejectsUntrustedClientCert() throws Exception {
+    int port = TestUtils.pickUnusedPort();
+
+    // Create & start a server. It requires client authentication and trusts only the test CA.
+    File serverCertFile = TestUtils.loadCert("localhost_server.pem");
+    File serverPrivateKeyFile = TestUtils.loadCert("localhost_server.key");
+    X509Certificate[] serverTrustedCaCerts = {
+      TestUtils.loadX509Cert("ca.pem")
+    };
+    Server server = serverBuilder(port, serverCertFile, serverPrivateKeyFile, serverTrustedCaCerts)
+        .addService(EchoServiceGrpc.bindService(new DummyEchoRpcService()))
+        .build()
+        .start();
+
+    try {
+      // Create a client. Its credentials come from a CA that the server does not trust. The client
+      // trusts both test CAs, so we can be sure that the handshake failure is due to the server
+      // rejecting the client's cert, not the client rejecting the server's cert.
+      File clientCertFile = TestUtils.loadCert("badclient.pem");
+      File clientPrivateKeyFile = TestUtils.loadCert("badclient.key");
+      X509Certificate[] clientTrustedCaCerts = {
+        TestUtils.loadX509Cert("ca.pem"),
+        TestUtils.loadX509Cert("badclient.pem")  // Cert is self-signed, and so is its own issuer.
```

Comment:
You are working around a non-existent bug? Couldn't that end up being its own downfall, if there is a different bug than you planned for? For example, what if the trustManager method was broken and dropped all but the first certificate? Then the test could pass because the client doesn't trust the server.

To me there is one way to go about it that has high confidence: make multiple tests and change only one thing between them, commonly one that passes and one that fails. I'm thinking along these lines:

1. no client certificate, server requires certificate (RPC fails)
2. untrusted client certificate, server requires certificate (RPC fails)
  * This has an additional line to set the cert on the client
3. trusted client certificate, server requires certificate (RPC succeeds)
  * This has an additional line to set the trusted CA on the server

It seems test 3, addresses the situation you mentioned. In my mind, the tests should be _minimal_ as well, just enough to get the RPC to fail/succeed. Since the client's CA isn't needed to be trusted on the client, don't trust it. That way, we are even more likely to find a bug or false-assumptions in the code. And if you have reason to remain uncomfortable, add a 4th test where the client trusts its own certs.